### PR TITLE
An alternative updater

### PR DESCRIPTION
**Pull request by kag0**

### CODE REVIEW

File: src/blackdoor/cqbe/node/Updater.java [hunk 1]
```diff
@@ -1,225 +1,244 @@
...
-			while (timer){
-				Thread.sleep(6000);
-				update();
+		
+		while(running){
+			try {
+				schedule();
+			} catch (InterruptedException e) {
+				DBP.printwarningln("Updater thread interrupted!");
+				run();
```
Comment (CjBuresch) [inline src/blackdoor/cqbe/node/Updater.java]: A recursive call when the thread has been interrupted? Calling stop() will throw the interrupt exeception, which will be caught here? Still kills the loop(s), given that running will be false. Am I missing something?

Comment (kag0) [inline src/blackdoor/cqbe/node/Updater.java]: The intention is to catch the interrupts not generated by stop(). That way the interrupt is caught, but running is still true, and the updater can continue.

File: src/blackdoor/cqbe/node/Updater.java [hunk 2]
```diff
@@ -1,225 +1,244 @@
...
+					}
+				}else{
+					try {
+						if(Router.ping(node)){
+							if(!Node.getAddressTable().contains(node))
+								DBP.printdemoln("Adding " + node.l3ToString() + " to table from updater");
+							Node.getAddressTable().add(node);
+							ref.forgive(node);
+						}else{
+							ref.strike(node);
```
Comment (ykravtsov) [inline src/blackdoor/cqbe/node/Updater.java]: why don't you combine ping and lookup by sending a lookup only to the nodes that respond to the ping

File: src/blackdoor/cqbe/node/Updater.java [hunk 3]
```diff
@@ -1,225 +1,244 @@
...
+		for(L3Address neighbor : Node.getAddressTable()){
+			try {
+				List<Address> keys = Router.getIndex(neighbor, 1);
+				for(Address key : keys){
+					if (!Node.getStorageController().containsKey(key)) {
+						byte[] value = Router.getValue(neighbor, key);
+						Node.getStorageController().put(
+								new CASFileAddress(Node.getStorageController()
+										.getDomain(), value));
+					}
```
Comment (ykravtsov) [inline src/blackdoor/cqbe/node/Updater.java]: Storage update needs to be multithreaded

File: src/blackdoor/cqbe/node/Updater.java (hunk 2 above)
Comment (ykravtsov) [inline src/blackdoor/cqbe/node/Updater.java]: why don't you combine ping and lookup by sending a lookup only to the nodes that respond to the ping

File: src/blackdoor/cqbe/node/Updater.java (hunk 3 above)
Comment (kag0) [inline src/blackdoor/cqbe/node/Updater.java]: yup

File: src/blackdoor/cqbe/node/Updater.java [hunk 4]
```diff
@@ -1,225 +1,244 @@
...
-				DBP.printdemoln("Adding " + c + " to address table from updater");
-				toAdd.add(c);
+		}
+		*/
+		//poll for new storage values
+		for(L3Address neighbor : Node.getAddressTable()){
+			try {
+				List<Address> keys = Router.getIndex(neighbor, 1);
+				for(Address key : keys){
+					if (!Node.getStorageController().containsKey(key)) {
```
Comment (kag0) [inline src/blackdoor/cqbe/node/Updater.java]: also need to check in this conditional if the key is one we want to store.

**Pull request merged**